pkg/cmd/trait_bind.go

```diff
@@ -118,51 +115,88 @@ func (o *commandOptions) Complete(cmd *cobra.Command, args []string, ctx context
-					updated = true
-					c.Traits[it] = corev1alpha2.ComponentTrait{Trait: runtime.RawExtension{Object: curObj}}
-					break
+func DetachTraitPlugins(parentCmd *cobra.Command, c client.Client, ioStreams cmdutil.IOStreams) error {
+	templates, err := plugins.GetTraitsFromCluster(context.TODO(), types.DefaultOAMNS, c)
+	if err != nil {
+		return err
+	}
+	ctx := context.Background()
+	for _, tmp := range templates {
+		var name = tmp.Alias
+		o := NewCommandOptions(ioStreams)
+		o.Client = c
+		o.Env, _ = GetEnv()
+		pluginCmd := &cobra.Command{
+			Use:                   name + ":detach <appname>",
+			DisableFlagsInUseLine: true,
+			Short:                 "Detach " + name + " trait from an app",
+			Long:                  "Detach " + name + " trait from an app",
+			Example:               `rudr scale:detach frontend`,
+			RunE: func(cmd *cobra.Command, args []string) error {
+				if err := o.DetachTrait(cmd, args, ctx); err != nil {
+					return err
 				}
-			}
-			if !updated {
-				c.Traits = append(c.Traits, corev1alpha2.ComponentTrait{Trait: runtime.RawExtension{Object: curObj}})
-			}
-			o.AppConfig.Spec.Components[ic] = c
-			break
+				return o.Run(cmd, ctx)
+			},
 		}
-	} else {
-		traitAlias := o.TraitAlias
-		_, _, tKind := cmdutil.GetTraitNameAliasKind(ctx, c, namespace, traitAlias)
-		//if tName == "" && tAlias == "" && tKind == "" {
-		//	errMsg := fmt.Sprintf("Error: trait name `%s` is NOT valid, please try again.", traitName)
-		//	cmdutil.PrintErrorMessage(errMsg, 1)
-		//}
-
-		traits := o.AppConfig.Spec.Components[0].Traits
-		traitDefinitionList := cmdutil.ListTraitDefinitionsByApplicationConfiguration(o.AppConfig)
-		for i := 0; i < len(o.AppConfig.Spec.Components[0].Traits); i++ {
-			if strings.EqualFold(traitDefinitionList[i].Kind, tKind) {
-				o.AppConfig.Spec.Components[0].Traits = append(traits[:i], traits[i+1:]...)
-				i--
+		pluginCmd.SetOut(o.Out)
+		o.TraitAlias = name
+		o.Detach = true
+		parentCmd.AddCommand(pluginCmd)
+	}
+	return nil
+}
+
+func (o *commandOptions) DetachTrait(cmd *cobra.Command, args []string, ctx context.Context) error {
+	argsLength := len(args)
+	if argsLength < 1 {
+		return errors.New("please specify the name of the app")
+	}
+	c := o.Client
+	namespace := o.Env.Namespace
+
+	var appName = args[0]
+	if err := c.Get(ctx, client.ObjectKey{Namespace: o.Env.Namespace, Name: appName}, &o.AppConfig); err != nil {
+		return err
+	}
+
+	_, _, tKind := cmdutil.GetTraitNameAliasKind(ctx, c, namespace, o.TraitAlias)
+
+	for _, com := range o.AppConfig.Spec.Components {
+		if com.ComponentName == appName {
+			traits := com.Traits
+			traitDefinitionList := cmdutil.ListTraitDefinitionsByApplicationConfiguration(o.AppConfig)
+			for i := 0; i < len(o.AppConfig.Spec.Components[0].Traits); i++ {
+				if strings.EqualFold(traitDefinitionList[i].Kind, tKind) {
+					o.AppConfig.Spec.Components[0].Traits = append(traits[:i], traits[i+1:]...)
+					i--
+				}
```

Comment:
It feels like we are hacking something here, why don't check one by one and append to a new list?

Reply:
Fixed per your suggestions